### PR TITLE
Closes #1763, Closes #1918 - Adding config for optionally importing archived paragraphs.

### LINKED ISSUE #1763: Content under moderation on source site doesn't migrate published paragraphs when latest revision is a DRAFT
## Problem/Motivation
Our paragraphs migrations do not migrate archived paragraphs.

Content moderation marks paragraphs as archived but renders them anyway if the current version is behind the lastest draft revision.

### To Reproduce
Create QS1 site
Turn on content moderation workflow
Add flexible page
add paragraphs to flexible page
Publish page
create new draft of page

Create QS2
Turn on content moderation workflow
Migrate source to destination
See paragraphs not migrated.

### PR DESCRIPTION
**Pull request by trackleft**
## Description
When migrating from sites with workbench_moderation enabled on pages with paragraphs attached, and the latest revision of a page is a draft, and all paragraphs are marked as archived, the archived paragraphs do not migrate and are not attached to the migrated page.

## Related issues
Closes #1763
Closes #1918

## How to test
On a UA Quickstart/Quickstart 1 source site ensure workbench_moderation is enabled and there is set up for uaqs_flexible_page.
On a Arizona Quickstart/Quickstart 2 destination site
Set the configuration `drush cset az_migration.settings allow_archived_paragraphs true`
On existing migrations: lando drush migrate:import --tag="Quickstart Paragraphs" --update
On new migrations: lando drush migrate:import --group="az_migration" 
Once you have imported all revisions, you should be able to rerun the flexible page migration.
lando drush migrate:import az_node_flexible_page --update

### Arizona Quickstart (install profile, custom modules, custom theme)
- Patch release changes
   - [x] Bug fix
   - [ ] Accessibility, performance, or security improvement
   - [ ] Critical institutional link or brand change
- Minor release changes
   - [x] New feature
   - [ ] Breaking or visual change to existing behavior
   - [ ] Non-critical brand change
   - [ ] New internal API or API improvement with backwards compatibility
   - [ ] Risky or disruptive cleanup to comply with coding standards
   - [ ] High-risk or disruptive change (requires upgrade path, risks regression, etc.)
- [ ] Other or unknown

### Drupal core
- Patch release changes
   - [ ] Security update
   - [ ] Patch level release (non-security bug-fix release)
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major or minor level update
- [ ] Other or unknown

### Drupal contrib projects
- Patch release changes
   - [ ] Security update
   - [ ] Patch or minor level update
   - [ ] Add new module
   - [ ] Patch removal that's no longer necessary
- Minor release changes
   - [ ] Major level update
- [ ] Other or unknown

## Checklist

- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 1]
```diff
@@ -29,8 +43,11 @@ public function query() {
         ])
       ->fields('pr', ['revision_id']);
     $query->innerJoin('paragraphs_item_revision', 'pr', static::JOIN);
-    // Omit archived (deleted) paragraphs.
-    $query->condition('p.archived', 0);
+
+    if (empty($this->sourceContentModerated) || $this->sourceContentModerated === 'false' || $this->sourceContentModerated === FALSE || $this->sourceContentModerated === 0) {
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:47]:
```suggestion
    if (empty($this->sourceContentModerated)) {
```

**Comment by trackleft** [comment]
Do more investigation into what exactly is happening on the source site, in order to refine the query in the source plugin, or at the very least add comments to the source plugin explaining what is happening.

File: modules/custom/az_migration/README.md [hunk 2]
```diff
@@ -30,6 +30,15 @@ In order to simplify the process of getting started with Quickstart migrations,
 
 To get started with a Quickstart 2 migration using the scaffolding repo as a project template, follow the steps in that repository's [README](https://github.com/az-digital/az-quickstart-scaffolding#readme) until you get to the "Migration setup in Lando" section and then follow these steps:
 
+_Note: If your source site has workbench moderation enabled and it is using it
+to create a workflow with any content type that has paragraphs(azqs_news,
```
Comment (akslay) [inline modules/custom/az_migration/README.md:34]:
```suggestion
to create a workflow with any content type that has paragraphs (azqs_news,
```

**Comment by trackleft** [comment]
Possibly investigate this https://git.drupalcode.org/project/drupal/-/blob/9.5.x/core/modules/node/src/Plugin/migrate/source/d7/NodeComplete.php
		https://www.drupal.org/node/3200949

File: modules/custom/az_migration/migrations/migrate_plus.migration.az_paragraph_contact.yml [hunk 3]
```diff
@@ -3,6 +3,7 @@ label: AZ Pragrapgh Contact
 migration_group: az_migration
```
Comment (bberndt-uaz) [inline modules/custom/az_migration/migrations/migrate_plus.migration.az_paragraph_contact.yml:3]: "AZ Pragrapgh Contact" can also be corrected on line 2 above.

File: modules/custom/az_migration/README.md [hunk 4]
```diff
@@ -30,6 +30,15 @@ In order to simplify the process of getting started with Quickstart migrations,
...
+azqs_flexible_page), you should use this command to allow archived paragraph
+revisions to be migrated. See https://github.com/az-digital/az_quickstart/issues/1763_
+
+```
+drush cset az_migration.settings az_source_content_moderated true
```
Comment (trackleft) [inline modules/custom/az_migration/README.md:39]: Make this more descriptive of what it actually does.
```suggestion
drush cset az_migration.settings az_migrate_allow_archived_paragraphs true
```

File: modules/custom/az_migration/config/install/az_migration.settings.yml [hunk 5]
```diff
@@ -3,3 +3,5 @@ migrate_d7_filebasepath: null
 migrate_d7_public_path: sites/default/files
 migrate_d7_private_path: null
 migrate_d7_temporary_path: tmp
+# If the source site is using content moderation, set az_source_content_moderated to true.
+az_source_content_moderated: false
```
Comment (trackleft) [inline modules/custom/az_migration/config/install/az_migration.settings.yml:7]:
```suggestion
az_migrate_allow_archived_paragraphs: false
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 6]
```diff
@@ -8,16 +8,30 @@
...
+   * @var bool
+   */
+  protected $sourceContentModerated;
+
   /**
    * {@inheritdoc}
    */
   public function query() {
+    // @phpstan-ignore-next-line
+    $this->sourceContentModerated = \Drupal::config('az_migration.settings')->get('az_source_content_moderated');
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:33]:
```suggestion
    $this->sourceContentModerated = \Drupal::config('az_migration.settings')->get('az_migrate_allow_archived_paragraphs');
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 7]
```diff
@@ -8,16 +8,30 @@
...
  * )
  */
 class AZParagraphsItem extends ParagraphsItem {
 
+  /**
+   * Whether the source content is under moderation.
+   *
+   * @var bool
+   */
+  protected $sourceContentModerated;
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:26]:
```suggestion
  protected $allowArchivedParagraphs;
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 8]
```diff
@@ -29,8 +43,11 @@ public function query() {
         ])
       ->fields('pr', ['revision_id']);
     $query->innerJoin('paragraphs_item_revision', 'pr', static::JOIN);
-    // Omit archived (deleted) paragraphs.
-    $query->condition('p.archived', 0);
+
+    if (empty($this->sourceContentModerated)) {
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:47]:
```suggestion
    if (empty($this->allowArchivedParagraphs)) {
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 9]
```diff
@@ -29,8 +43,11 @@ public function query() {
         ])
       ->fields('pr', ['revision_id']);
     $query->innerJoin('paragraphs_item_revision', 'pr', static::JOIN);
-    // Omit archived (deleted) paragraphs.
-    $query->condition('p.archived', 0);
+
+    if (empty($this->allowArchivedParagraphs)) {
+      // Omit archived (deleted) paragraphs.
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:48]:
```suggestion
      // Omit archived (deleted or stale) paragraphs.
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 10]
```diff
@@ -29,8 +43,11 @@ public function query() {
         ])
       ->fields('pr', ['revision_id']);
     $query->innerJoin('paragraphs_item_revision', 'pr', static::JOIN);
-    // Omit archived (deleted) paragraphs.
-    $query->condition('p.archived', 0);
+
+    if (empty($this->allowArchivedParagraphs)) {
```
Comment (joeparsons) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:47]: I think this is wrong.  `empty(FALSE)` would return `TRUE`.
```suggestion
    if (!empty($this->allowArchivedParagraphs)) {
```

Comment (joeparsons) [reply modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:47]: Nevermind.  I think this is OK.  `empty()` is just confusing here.

File: modules/custom/az_migration/config/install/az_migration.settings.yml [hunk 11]
```diff
@@ -3,3 +3,5 @@ migrate_d7_filebasepath: null
 migrate_d7_public_path: sites/default/files
 migrate_d7_private_path: null
 migrate_d7_temporary_path: tmp
+# If the source site is using content moderation, set az_source_content_moderated to true.
+az_migrate_allow_archived_paragraphs: false
```
Comment (trackleft) [inline modules/custom/az_migration/config/install/az_migration.settings.yml:7]:
```suggestion
allow_archived_paragraphs: false
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 12]
```diff
@@ -8,16 +8,30 @@
...
+   * @var bool
+   */
+  protected $allowArchivedParagraphs;
+
   /**
    * {@inheritdoc}
    */
   public function query() {
+    // @phpstan-ignore-next-line
+    $this->allowArchivedParagraphs = \Drupal::config('az_migration.settings')->get('az_migrate_allow_archived_paragraphs');
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:33]:
```suggestion
    $this->allowArchivedParagraphs = \Drupal::config('az_migration.settings')->get('allow_archived_paragraphs');
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php (hunk 10 above)
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:47]:
```suggestion
    if (!$this->allowArchivedParagraphs) {
```

File: modules/custom/az_migration/README.md [hunk 13]
```diff
@@ -30,6 +30,15 @@ In order to simplify the process of getting started with Quickstart migrations,
 
 To get started with a Quickstart 2 migration using the scaffolding repo as a project template, follow the steps in that repository's [README](https://github.com/az-digital/az-quickstart-scaffolding#readme) until you get to the "Migration setup in Lando" section and then follow these steps:
 
+_Note: If your source site has workbench moderation enabled and it is using it
+to create a workflow with any content type that has paragraphs (azqs_news,
+azqs_flexible_page), you should use this command to allow archived paragraph
+revisions to be migrated. See https://github.com/az-digital/az_quickstart/issues/1763_
+
+```
+drush cset az_migration.settings az_migrate_allow_archived_paragraphs true
```
Comment (trackleft) [inline modules/custom/az_migration/README.md:39]:
```suggestion
drush cset az_migration.settings allow_archived_paragraphs true
```

File: modules/custom/az_migration/config/install/az_migration.settings.yml [hunk 14]
```diff
@@ -3,3 +3,5 @@ migrate_d7_filebasepath: null
 migrate_d7_public_path: sites/default/files
 migrate_d7_private_path: null
 migrate_d7_temporary_path: tmp
+# If the source site is using content moderation, set az_source_content_moderated to true.
```
Comment (trackleft) [inline modules/custom/az_migration/config/install/az_migration.settings.yml:6]:
```suggestion
# If the source site is using content moderation, set allow_archived_paragraphs to true.
```

File: modules/custom/az_migration/config/install/az_migration.settings.yml [hunk 15]
```diff
@@ -3,3 +3,5 @@ migrate_d7_filebasepath: null
 migrate_d7_public_path: sites/default/files
 migrate_d7_private_path: null
 migrate_d7_temporary_path: tmp
+# If the source site is using content moderation, set allow_archived_paragraphs to true.
```
Comment (trackleft) [inline modules/custom/az_migration/config/install/az_migration.settings.yml:6]:
```suggestion
# If the source site is using workbench moderation, set allow_archived_paragraphs to true.
```

File: modules/custom/az_migration/config/install/az_migration.settings.yml [hunk 16]
```diff
@@ -3,3 +3,5 @@ migrate_d7_filebasepath: null
 migrate_d7_public_path: sites/default/files
 migrate_d7_private_path: null
 migrate_d7_temporary_path: tmp
+# If the source site is using workbench moderation, set allow_archived_paragraphs to true.
```
Comment (trackleft) [inline modules/custom/az_migration/config/install/az_migration.settings.yml:6]:
```suggestion
# If the source site is using the workbench_moderation module to add workflows on nodes with paragraphs on them, set allow_archived_paragraphs to true.
```

File: modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php [hunk 17]
```diff
@@ -8,16 +8,30 @@
...
+ *   of that particular type.
+ *
  * @MigrateSource(
  *   id = "az_paragraphs_item"
  * )
  */
 class AZParagraphsItem extends ParagraphsItem {
 
+  /**
+   * Whether the source content is under moderation.
```
Comment (trackleft) [inline modules/custom/az_migration/src/Plugin/migrate/source/AZParagraphsItem.php:22]:
```suggestion
   * Whether to migrate archived paragraphs.
```

**Comment by joeparsons** [comment]
Worth noting that this includes changes that effectively revert the workaround we added to allow us to update Paragraphs to 8.x-1.14 without breaking our migrations (#1545).

**Review by joeparsons — APPROVED** [review]

**Review by jwmoore1 — APPROVED** [review]

**Pull request merged**